packages/contracts-bedrock/test/setup/Upgrade.s.sol

```diff
@@ -0,0 +1,114 @@
+
+// Libraries
+import { Constants } from "src/libraries/Constants.sol";
+import { GameTypes } from "src/dispute/lib/Types.sol";
+import { EIP1967Helper } from "test/mocks/EIP1967Helper.sol";
+
+// Interfaces
+import { IFaultDisputeGame } from "interfaces/dispute/IFaultDisputeGame.sol";
+import { IDisputeGame } from "interfaces/dispute/IDisputeGame.sol";
+import { IDisputeGameFactory } from "interfaces/dispute/IDisputeGameFactory.sol";
+import { IAddressManager } from "interfaces/legacy/IAddressManager.sol";
+
+/// @title Upgrade
+/// @notice This script is called by Setup.sol as a preparation step for the foundry test suite, and is run as an
+///         alternative to Deploy.s.sol, when `UPGRADE_TEST=true` is set in the env.
+///         Like Deploy.s.sol this script saves the system addresses to disk so that they can be read into memory later
+///         on, however rather than deploying new contracts from the local source code, it simply reads the addresses
+///         from the superchain-registry.
+///         Therefore this script can only be run against a fork of a production network which is listed in the
+///         superchain-registry.
+///         This contract must not have constructor logic because it is set into state using `etch`.
+contract Upgrade is Deployer {
+    using stdJson for string;
+
+    /// @notice Returns the base chain name to use for forking
+    /// @return The base chain name as a string
+    function upgradeBaseChain() internal view returns (string memory) {
+        return vm.envOr("UPGRADE_BASE_CHAIN", string("mainnet"));
+    }
+
+    /// @notice Returns the OP chain name to use for forking
+    /// @return The OP chain name as a string
+    function upgradeOpChain() internal view returns (string memory) {
+        return vm.envOr("UPGRADE_OP_CHAIN", string("op"));
+    }
+
+    /// @notice Reads a standard chains addresses from the superchain-registry and saves them to disk.
+    function run() public {
+        string memory superchainBasePath = "./lib/superchain-registry/superchain/configs/";
+
+        // Read the superchain config files
+        // During development of an upgrade which adds a new contract, the contract will not yet be present in the
+        // superchain-registry. In this case, the contract will be deployed by the upgrade process, and will need to
+        // be saved by after the call to opcm.upgrade().
+        // After the upgrade is complete, the superchain-registry will be updated and the contract will be present.
+        // At this point, the test will need to be updated to read the new contract from the superchain-registry using
+        // either the `saveProxyAndImpl` or `save` functions.
+        string memory superchainToml =
+            vm.readFile(string.concat(superchainBasePath, upgradeBaseChain(), "/superchain.toml"));
+        string memory opToml =
+            vm.readFile(string.concat(superchainBasePath, upgradeBaseChain(), "/", upgradeOpChain(), ".toml"));
```

Comment:
We may be able to simplify things by collapsing `UPGRADE_BASE_CHAIN` and `UPGRADE_OP_CHAIN` into `OP_CHAIN_ID` and reading all addresses from https://github.com/ethereum-optimism/superchain-registry/blob/main/superchain/extra/addresses/addresses.json

That is similar to what we are doing in superchain-ops: https://github.com/ethereum-optimism/superchain-ops/pull/406

Reply:
I think it would be a significant refactor to this code path to:
- replace toml with json parsing
- needing to infer sepolia/mainnet from the data in the json file

And I'm not sure it gets us all that much. If this turns out not to scale well we can fix in the future. 
I'd like to resolve this without tracking it in an issue, please let me know if you feel strongly to the contrary.